traffic_ops/traffic_ops_golang/monitoring/monitoring.go

```diff
@@ -188,7 +188,7 @@ SELECT
 	status.name as status,
 	cachegroup.name as cachegroup,
 	me.tcp_port as port,
-	profile.name as profile,
+	(SELECT STRING_AGG(sp.profile_name, '+' ORDER by sp.priority ASC) FROM server_profile AS sp where sp.server=me.id group by sp.server) as profile,
```

Comment:
This aggregation will break if I have a Profile that contains `+` in its Name.

Why build up this big simulated Profile-aggregate-pseudo-name thingy instead of just getting an array of Profile names? Seems like that would be easier to deal with than joining and splitting on a special string.

Reply:
As an example( in monitoring.json), I have a traffic server as follows:
```
"trafficServers": [{
	**"profile": "to+global"**,
	"status": "ONLINE",
	"port": 80,
	"cachegroup": "mb",
	"hostname": "test2",
	"fqdn": "test2.abc.com",
	"interfaces": [{
		"ipAddresses": [{
			"address": "3.4.5.6",
			"gateway": null,
			"serviceAddress": true
		}],
		"maxBandwidth": null,
		"monitor": true,
		"mtu": null,
		"name": "bond0"
	}],
	"type": "EDGE",
	"hashid": "e5740dc9-970e-4016-85fe-5cd9d2b80a8e",
	"deliveryServices": [{
		"xmlId": "bat"
	}]
```
Note the profile, it is a layered profile.. So, it can be seen as an array of profiles`["to, "global"]` or shows as an agg of profiles(`to+global`).

So, now when you go to profile section in monitoring.json, you see the above profile: `to+global` as
```
{
	"name": "to+global",
	"type": "EDGE",
	"parameters": {
		"health.threshold.loadavg": 88,
		"health.threshold.queryTime": 1500,
		"history.count": 30
}
```
and makes it more convenient to read and understand.

Reply:
Also, it doesn't affect existing monitoring.json since Profile field in traffic servers remains unchanged.. It is still a string. Reason I mention this is because I am not a 100% sure how TMs use monitoring.json and whether we have to change TM to now start accepting an array field instead of string.

Reply:
Also, it doesn't affect existing monitoring.json since Profile field in traffic servers remains unchanged.. It is still a string. Reason I mention this is because I am not a 100% sure how TMs use monitoring.json and whether we have to change TM to now start accepting an array field instead of string.

Reply:
I don't know that just making TM understand profile layering would result in any more duplication than copying different arrangements of profiles into their own pseudo profiles. It seems like the two best options are

1. Calculate the needed Parameters for each cache server in TO and store them on a per-cache basis without making TM look up the name of the Profile to find the Parameters. This follows the pattern of making the configuration simple for TM to understand and not requiring it to build out relations through extrapolation. However, it probably makes monitoring configurations much bigger.
2. Tell TM what Profiles each server has and what Parameters belong to the relevant Profiles. This is in keeping with how things work today, just makes it a little more work to figure out a cache server's Parameters. I think the extra time this would take is probably negligible for any reasonable use-case. It would also cut down on duplication.

Personally I think option 2 makes the most sense, but in either case it'll require TM changes. At any rate, this "+" stuff won't work because if I have the profiles "a", "b", "c", "a+b", and "b+c", it will be literally impossible to tell what the entry "a+b+c" means in the monitoring payload. More to the point, keys must also be unique, so any time you have a profile with a name that can be constructed through "+" concatenation of any number of other profile names, you wind up with a silently broken configuration.

Reply:
I think the main reason to generate the "pseudo-aggregate" profile is to avoid breaking the monitoring config format unnecessarily. If we do that, we have to worry about things like keeping 4.0 backwards-compatible while letting 5.0 break, which means more version-handling code in TO and TM. If we just keep it fully backwards-compatible by using this "pseudo-aggregate" profile approach, then we don't need any of that version-handling code at all. Plus, TM wouldn't need to change at all with this approach, because TO has already handled that "profile aggregation" for us.

If we're worried that joining profile names with `+` to create the pseudo-aggregate profile name would cause problems, then we could choose a much less-likely set of characters to join with -- e.g. `+_+_+` or something else I would think is highly unlikely of being used in a real profile name.

Reply:
The only character not allowed in Profile names is a space, so I'd recommend that. But I still don't really like the idea. There's already separate handling TM for different API versions of the monitoring configuration, so it really ought not to be a big deal to handle it appropriately. Like as a backward compatibility thing for existing API versions, this is a good idea, but I really think APIv5 should just show you the actual data and TM can parse appropriately.

Reply:
From a troubleshooting perspective, seeing the pseudo-aggregate profiles in the monitoring snapshot is more helpful, because you don't have to do the aggregation in your head when looking at it to see what actual health thresholds are being used by TM for a given server. I'm trying to think of `monitoring.json` as more of a config file than an API endpoint. From an API perspective, sure, it would be nice to have that conformity w/ the servers API, but from a config file perspective, breaking the format doesn't really provide any real benefit to TM here and makes the config more confusing to operators. TM doesn't actually gain any kind of additional functionality through this breaking change, so why break it?

If this same change happened to be necessary for TR and the CRConfig, my opinion would be the same, but thankfully the CRConfig doesn't list individual profiles/params like monitoring.json. In general, the CRConfig has remained backwards-compatible even though the API as a whole has undergone many breaking changes. I'd think monitoring.json could be the same -- jumping through API version hoops when upgrading for no tangible benefit is a bit of a pain.

Reply:
What would be *most* helpful to TM would be to not have to look up profiles at all, and just get the data it needs on each server without needing to know what Profiles and Parameters even are.

Reply:
Sure, that might shave off a handful of code from TM, but it still adds backwards-compatibility issues that humans need to worry about.